adr/ADR-33.md

```diff
@@ -0,0 +1,257 @@
+# Service API
+
+| Metadata | Value                 |
+| -------- | --------------------- |
+| Date     | 2022-11-23            |
+| Author   | @aricart              |
+| Status   | Partially Implemented |
+| Tags     | client                |
+
+## Context and Problem Statement
+
+Simplify the development of NATS micro-services.
+
+The design goal of the API is to reduce development to having a complexity
+similar to that of writing a NATS subscription, but then by using simple
+configuration, allow the specified metadata to allow for standardization of
+discovery and observability.
+
+## Design
+
+Service configuration relies on the following:
+
+- `name` - really the _kind_ of the service. Shared by all the services that
+  have the same name. This `name` can only have
+  `A-Z, a-z, 0-9, dash, underscore`.
+- `version` - a SemVer string - impl should validate that this is SemVer
+- `description` - a human-readable description about the service (optional)
+- `schema`: (optional)
+  - `request` - a string/url describing the format of the request payload can be
+    JSON schema etc.
+  - `response` - a string/url describing the format of the response payload can
+    be JSON schema etc.
+- `statsHandler` - an optional function that returns unknown data that can be
+  serialized as JSON. The handler will be provided the endpoint for which it is
+  building a `EndpointStats`
+- `endpoint` - a subject and a handler (effectively equivalent to a NATS
+  subscription)
+
+All services are created using a function called `addService()` where the above
+options are passed. The function returns an object/struct that represents the
+service. At a minimum the service is expected to offer functions/methods that
+allow:
+
+- a `stop(error?)` function that allows user code to stop the service.
+  Optionally this function should allow for an optional error. Stop should
+  always drain its service subscriptions.
+- `reset()` to reset any tracked metrics
+- `info()` returns the [`ServiceInfo`](#INFO)
+- `stats()` to return the stats of the service
+- A callback handler or promise where the framework can notify when the service
+  has stopped. Note that this is independent of the NATS connection, and it
+  should be possible to run multiple services under a single connection.
+
+On startup a service is assigned an unique `id`. This `id` is used to
+distinguish different instances of the service and allow for a specific instance
+of the service to be addressed.
+
+### Discovery and Status
+
+Using the specified `name` and automatically generated `id` the service will
+automatically create a subscription to handle discovery and monitoring requests.
+
+The subject for discovery and requests is always composed of all capital
+letters. Prefixed by `$SRV`. Note that this prefix needs to be overridable much
+in the way as we do for `$JS`, in order to enable targetting tools to work
```

Comment:
Why? My objections to this are:

* The service name specification allows lower case letters. This throws that sensitivity away. A service named `MyFooService` will clash with a service called `MYfooSERVICE`, both will make a subject like `$SRV.PING.MYFOOSERVICE`

* Services are given id's based on NUID. NUIDS are mixed case. So instead of 
`$SRV.PING.MyFooService.iXIILpk27Y8CZ4iDP9KzJc` the id will be modified so `$SRV.PING.MYFOOSERVICE.IXIILPK27Y8CZ4IDP9KZJC` which is a different value.

* It's just not necessary to uppercase the name and id portion.

Reply:
One option would be to require only uppercase in the name. Also we should specify that service id is uppercase.